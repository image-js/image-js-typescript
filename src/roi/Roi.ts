import { Mask } from '../Mask';

import { RoiMap } from './RoiMapManager';
import { getMask } from './getMask';

export class Roi {
  private map: RoiMap;
  public id: number;
  public row: number;
  public column: number;
  public width: number;
  public height: number;
  public surface: number;

  public constructor(map: RoiMap, id: number) {
    this.map = map;
    this.id = id;
    this.row = map.height;
    this.column = map.width;
    this.width = 0;
    this.height = 0;
    this.surface = 0;
  }

  /**
   * Get the ROI map of the original image.
   *
   * @returns The ROI map.
   */
  public getMap(): RoiMap {
    return this.map;
  }

  /**
   * Return the value at the given coordinates in an ROI map.
   *
   * @param column - Column of the value.
   * @param row - Row of the value.
   * @returns The value at the given coordinates.
   */
  public getMapValue(column: number, row: number) {
    return this.map.data[this.map.width * row + column];
  }

  /**
   * Return the ratio between the width and the height of the bounding rectangle of the ROI.
   *
   * @returns The width by height ratio.
   */
  public getRatio(): number {
    return this.width / this.height;
  }

  /**
   * Generate a mask the size of the bounding rectangle of the ROI, where the pixels inside the ROI are set to true and the rest to false.
   *
   * @returns The ROI mask.
   */
  public getMask(): Mask {
    return getMask(this);
  }
}
