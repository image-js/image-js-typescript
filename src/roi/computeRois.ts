import { Roi } from './Roi';
import { RoiMapManager } from './RoiMapManager';

/**
 * Generate an array of ROIs based on an ROI map manager.
 *
 * @param roiMapManager - Roi map manager to use
 */
export function computeRois(roiMapManager: RoiMapManager): void {
  const map = roiMapManager.getMap();

  const whites = new Array(map.nbPositive);
  const blacks = new Array(map.nbNegative);

  for (let i = 0; i < map.nbPositive; i++) {
    whites[i] = {
      minRow: map.height,
      minColumn: map.width,
      maxRow: -1,
      maxColumn: -1,
      surface: 0,
      id: i + 1,
    };
  }
  for (let i = 0; i < map.nbNegative; i++) {
    blacks[i] = {
      minRow: map.height,
      minColumn: map.width,
      maxRow: -1,
      maxColumn: -1,
      surface: 0,
      id: -i - 1,
    };
  }

  for (let row = 0; row < map.height; row++) {
    for (let column = 0; column < map.width; column++) {
      let currentIndex = roiMapManager.getMapValue(row, column);

      let currentRoi;
      if (currentIndex < 0) {
        currentRoi = blacks[-currentIndex - 1];
      } else {
        currentRoi = whites[currentIndex - 1];
      }

      currentRoi.surface++;

      if (row < currentRoi.minRow) {
        currentRoi.minRow = row;
      }
      if (row > currentRoi.maxRow) {
        currentRoi.maxRow = row;
      }
      if (column < currentRoi.minColumn) {
        currentRoi.minColumn = column;
      }
      if (column > currentRoi.maxColumn) {
        currentRoi.maxColumn = column;
      }
    }
  }

  roiMapManager.whiteRois = new Array(map.nbPositive);
  roiMapManager.blackRois = new Array(map.nbNegative);

  for (let i = 0; i < map.nbPositive; i++) {
    let whiteRoi = new Roi(map, i);
    whiteRoi.row = whites[i].minRow;
    whiteRoi.column = whites[i].minColumn;
    whiteRoi.width = whites[i].maxColumn - whites[i].minColumn + 1;
    whiteRoi.height = whites[i].maxRow - whites[i].minRow + 1;
    whiteRoi.surface = whites[i].surface;
    whiteRoi.id = whites[i].id;

    roiMapManager.whiteRois[i] = whiteRoi;
  }
  for (let i = 0; i < map.nbNegative; i++) {
    let blackRoi = new Roi(map, i);
    blackRoi.row = blacks[i].minRow;
    blackRoi.column = blacks[i].minColumn;
    blackRoi.width = blacks[i].maxColumn - blacks[i].minColumn + 1;
    blackRoi.height = blacks[i].maxRow - blacks[i].minRow + 1;
    blackRoi.surface = blacks[i].surface;
    blackRoi.id = blacks[i].id;

    roiMapManager.blackRois[i] = blackRoi;
  }
}
