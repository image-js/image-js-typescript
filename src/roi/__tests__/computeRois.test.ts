import { fromMask } from '..';
import { Roi } from '../Roi';
import { computeRois } from '../computeRois';

describe('computeRois', () => {
  it('1x2 mask', () => {
    const mask = testUtils.createMask([[1, 0]]);
    const roiMapManager = fromMask(mask);
    computeRois(roiMapManager);

    const whiteRoi = new Roi(roiMapManager.getMap(), 1);
    whiteRoi.row = 0;
    whiteRoi.column = 0;
    whiteRoi.height = 1;
    whiteRoi.width = 1;
    whiteRoi.surface = 1;

    expect(roiMapManager.whiteRois).toHaveLength(1);
    expect(roiMapManager.whiteRois).toStrictEqual([whiteRoi]);

    const blackRoi = new Roi(roiMapManager.getMap(), -1);
    blackRoi.row = 0;
    blackRoi.column = 1;
    blackRoi.height = 1;
    blackRoi.width = 1;
    blackRoi.surface = 1;

    expect(roiMapManager.blackRois).toHaveLength(1);
    expect(roiMapManager.blackRois).toStrictEqual([blackRoi]);
  });
  it('3x3 mask', () => {
    const mask = testUtils.createMask([
      [1, 1, 1],
      [1, 0, 0],
      [0, 0, 1],
    ]);
    const roiMapManager = fromMask(mask);
    computeRois(roiMapManager);
    expect(roiMapManager.whiteRois).toHaveLength(2);
    expect(roiMapManager.blackRois).toHaveLength(1);

    expect(roiMapManager.whiteRois).toMatchSnapshot();
    expect(roiMapManager.blackRois).toMatchSnapshot();
  });
});
