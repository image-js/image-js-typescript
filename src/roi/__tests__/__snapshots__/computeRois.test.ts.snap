// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`computeRois 3x3 mask 1`] = `
Array [
  Roi {
    "column": 0,
    "height": 2,
    "id": 1,
    "map": Object {
      "data": Int16Array [
        1,
        1,
        1,
        1,
        -1,
        -1,
        -1,
        -1,
        2,
      ],
      "height": 3,
      "nbNegative": 1,
      "nbPositive": 2,
      "width": 3,
    },
    "row": 0,
    "surface": 4,
    "width": 3,
  },
  Roi {
    "column": 2,
    "height": 1,
    "id": 2,
    "map": Object {
      "data": Int16Array [
        1,
        1,
        1,
        1,
        -1,
        -1,
        -1,
        -1,
        2,
      ],
      "height": 3,
      "nbNegative": 1,
      "nbPositive": 2,
      "width": 3,
    },
    "row": 2,
    "surface": 1,
    "width": 1,
  },
]
`;

exports[`computeRois 3x3 mask 2`] = `
Array [
  Roi {
    "column": 0,
    "height": 2,
    "id": -1,
    "map": Object {
      "data": Int16Array [
        1,
        1,
        1,
        1,
        -1,
        -1,
        -1,
        -1,
        2,
      ],
      "height": 3,
      "nbNegative": 1,
      "nbPositive": 2,
      "width": 3,
    },
    "row": 1,
    "surface": 4,
    "width": 3,
  },
]
`;
